### ISSUE TITLE
 Phalcon 2.0 not returning DISTINCT column in PHQL

### ISSUE THREAD

**Issue by alexetnunes** [phalcon/cphalcon#9985]
After upgrading to Phalcon 2.0 from 1.3.4 any PHQL queries using the DISTINCT syntax would return a blank column on whatever column had distinct.

Example:
![screen shot 2015-04-08 at 3 19 04 pm](https://cloud.githubusercontent.com/assets/6401210/7052074/9e37d18c-de02-11e4-82b5-07fba714439b.png)

The Robots.model column will return blank but everything else would comeback fine.

Is there an issue around the DISTINCT syntax in PHQL, or is not used properly?

OS: Mac OSX 10.10.2 
MAMP 3.1
PHP 5.6.6
Phalcon 2.0.0-RC1

**Comment by andresgutierrez** [comment]
I'll take a look

**Comment by andresgutierrez** [comment]
I can't reproduce this, I've the following test and it's producing the right SQL:

```php
<?php

use Phalcon\DI,
    Phalcon\Db\Column,
    Phalcon\Events\Manager as EventsManager,
    Phalcon\Db\Adapter\Pdo\Mysql as Connection,
    Phalcon\Mvc\Model\Manager as ModelsManager,
    Phalcon\Mvc\Model\Metadata\Memory as ModelsMetaData;

$eventsManager = new EventsManager();

$di = new DI();

$connection = new Connection(array(
    "host"     => "localhost",
    "username" => "root",
    "password" => "",
    "dbname"   => "phalcon_test"
));

$connection->setEventsManager($eventsManager);

$eventsManager->attach('db',
    function ($event, $connection) {
        switch ($event->getType()) {
            case 'beforeQuery':
                echo $connection->getSqlStatement(), "\n";
                break;
        }
    }
);

$modelsManager = new ModelsManager();
$modelsManager->setDi($di);
$di['db'] = $connection;
$di['modelsManager'] = $modelsManager;
$di['modelsMetadata'] = new ModelsMetadata();

class Robots extends \Phalcon\Mvc\Model
{
}

$modelsManager->executeQuery('SELECT DISTINCT name, Robots.* FROM Robots');
```

```sql
SELECT DISTINCT `robots`.`name` AS `_0`, `robots`.`id` AS `_robots_id`, `robots`.`name` AS `_robots_name`, `robots`.`type` AS `_robots_type`, `robots`.`year` AS `_robots_year`, `robots`.`datetime` AS `_robots_datetime`, `robots`.`text` AS `_robots_text` FROM `robots`
```

could you please provide more information?

**Comment by alexetnunes** [comment]
The only difference I can see is that I have a left join on the PHQL query. Maybe the error happens with a join. If you could try it with the "LEFT JOIN" syntax adding  a table that would be appreciated.

**Comment by alexetnunes** [comment]
Update: I figured out the issue I was having, after taking a close look.

The query was returning the distinct column at the numerical index rather than the associative index, the column name. So when I was trying to access the column in Volt it would return blank. I added the AS "column name to the DISTINCT column and everything worked fine.

Just as confirmation in 1.3.4 you can do this:
`$modelsManager->executeQuery('SELECT DISTINCT name, Robots.* FROM Robots');`

But in Phalcon 2 you need the AS, see below:
`$modelsManager->executeQuery('SELECT DISTINCT name AS distinct_name, Robots.* FROM Robots');`

Which will return that column under the name distinct_name for Volt.

Please let me know if this is right. And thank you in advance for your help.

**Comment by dugwood** [comment]
I've just upgraded to the latest 2.0.0, and I can confirm the behaviour: it was working great with a previous 2.0.0 release, and now, if I add `->distinct(true)` to the query builder, I don't get any results.

I'm looking for the issue in the code right now.

**Comment by dugwood** [comment]
@alexetnunes I can confirm you the behaviour of the needed `AS`: I was selecting only one column, and there was no result at all, same thing with 2 columns. But if I add a `AS`, I get the expected result.

**Comment by dugwood** [comment]
@andresgutierrez the SQL is good, I dumped it and it is valid SQL statement. In fact, that's the only part that hasn't changed. For example:
`SELECT DISTINCT country FROM Users WHERE user_status = 1` => the SQL query is the same as before, and it's working when tested in the database (with the expected result).

**Comment by dugwood** [comment]
@andresgutierrez, sorry, but I didn't find out what the issue was. Maybe it's a complex vs simple resultset issue (as when I add the `AS` keyword, I'm receiving a SimpleResultset instead of an empty ComplexResultset). For now, the solution by @alexetnunes is working for me: `->columns('country AS country')`.
But I agree it's a bug.

**Comment by andresgutierrez** [comment]
I still can't reproduce this in 2.0.3. Can you please try again using this version?

**Comment by dugwood** [comment]
@andresgutierrez, I can't reproduce the bug either! Really strange... I've just tested it against 2.0.x and 2.0.0, and it's working fine (with or without the column's alias in SQL).

I think I've tested it against the master at the time, so the issue may be resolved as of now.

@alexetnunes can you confirm it too?

@andresgutierrez : where can I find the 2.0.3 stable release in github? I can't find any branch or tag that match the latest stable release.

**Comment by sergeyklay** [comment]
@dugwood 2.0.x == 2.0.3, after 2.0.3 release 2.0.x == 2.0.4, etc

**Comment by alexetnunes** [comment]
This was fixed for sure as of 2.0.1, since then I have been able to use it without an alias. 

It was included in the release notes: http://blog.phalconphp.com/post/118465365385/phalcon-2-0-1-released  as "Added support for general SELECT ALL/SELECT DISTINCT in PHQL" 

Thanks for the support!
